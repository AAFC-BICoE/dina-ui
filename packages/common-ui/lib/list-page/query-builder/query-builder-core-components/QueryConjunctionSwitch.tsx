import { DinaMessage } from "../../../../../dina-ui/intl/dina-ui-intl";
import Button from "react-bootstrap/Button";

interface QueryConjunctionSwitchProps {
  currentConjunction?: string;
  setConjunction?: (conjunction: string) => void;
}

export function QueryConjunctionSwitch({
  currentConjunction,
  setConjunction
}: QueryConjunctionSwitchProps) {
  return (
    <>
      <div className="toggleGroup">
        <Button
          className={
            currentConjunction === "AND"
              ? "toggleButton activeToggle"
              : "toggleButton"
          }
          onClick={(_) => setConjunction?.("AND")}
        >
          <DinaMessage id="queryBuilder_conjunction_and" />
        </Button>
        <Button
          className={
            currentConjunction === "OR"
              ? "toggleButton activeToggle"
              : "toggleButton"
          }
          onClick={(_) => setConjunction?.("OR")}
        >
          <DinaMessage id="queryBuilder_conjunction_or" />
        </Button>
      </div>
    </>
  );
}
