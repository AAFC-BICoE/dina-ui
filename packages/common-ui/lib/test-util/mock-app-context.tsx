import { mount } from "enzyme";
import { merge, noop } from "lodash";
import { useMemo, useRef } from "react";
import { SWRConfig } from "swr";
import { PartialDeep } from "type-fest";
import { AccountContextI, AccountProvider } from "../account/AccountProvider";
import {
  ApiClientI,
  ApiClientImpl,
  ApiClientProvider
} from "../api-client/ApiClientContext";
import { CommonUIIntlProvider } from "../intl/common-ui-intl";
import { ModalProvider } from "../modal/modal";

interface MockAppContextProviderProps {
  apiContext?: PartialDeep<ApiClientI>;
  accountContext?: Partial<AccountContextI>;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  accountContext,
  apiContext = { apiClient: { get: () => undefined as any } },
  children
}: MockAppContextProviderProps) {
  const DEFAULT_MOCK_ACCOUNT_CONTEXT: AccountContextI = useMemo(
    () => ({
      authenticated: true,
      groupNames: ["aafc", "cnc"],
      initialized: true,
      login: noop,
      logout: noop,
      roles: ["user"],
      token: "test-token",
      username: "test-user",
      isAdmin: false
    }),
    []
  );

  const DEFAULT_API_CONTEXT_VALUE = useMemo(
    () =>
      new ApiClientImpl({
        newId: () => "00000000-0000-0000-0000-000000000000"
      }),
    []
  );

  const apiContextWithWarnings = {
    ...apiContext,
    // Add a warning when bulkGet doesn't return anything in a test:
    bulkGet: async (paths: string[], options) => {
      const resources = await apiContext?.bulkGet?.(paths, options);
      if (!resources) {
        console.warn("No response returned for bulkGet paths: ", paths);
      } else {
        resources.forEach((resource, index) => {
          if (!resource) {
            console.warn("No value returned for bulkGet path: ", paths[index]);
          }
        });
      }
      return resources;
    }
  };

  const modalWrapperRef = useRef<HTMLDivElement>(null);

  return (
    <SWRConfig
      // Reset SWR cache between tests.
      value={{ provider: () => new Map() }}
    >
      <AccountProvider
        value={{ ...DEFAULT_MOCK_ACCOUNT_CONTEXT, ...accountContext }}
      >
        <ApiClientProvider
          value={merge({}, DEFAULT_API_CONTEXT_VALUE, apiContextWithWarnings)}
        >
          <CommonUIIntlProvider>
            <div ref={modalWrapperRef}>
              <ModalProvider appElement={modalWrapperRef.current}>
                {children}
              </ModalProvider>
            </div>
          </CommonUIIntlProvider>
        </ApiClientProvider>
      </AccountProvider>
    </SWRConfig>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}
