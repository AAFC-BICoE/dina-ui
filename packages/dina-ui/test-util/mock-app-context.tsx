import {
  AccountContextI,
  AccountProvider,
  ApiClientI,
  ApiClientImpl,
  ApiClientProvider,
  ModalProvider
} from "common-ui";
import { mount } from "enzyme";
import { merge, noop } from "lodash";
import { useMemo, useRef } from "react";
import { DndProvider } from "react-dnd-cjs";
import HTML5Backend from "react-dnd-html5-backend-cjs";
import { SWRConfig } from "swr";
import { PartialDeep } from "type-fest";
import { FileUploadProviderImpl } from "../components/object-store/file-upload/FileUploadProvider";
import { DinaIntlProvider } from "../intl/dina-ui-intl";

interface MockAppContextProviderProps {
  apiContext?: PartialDeep<ApiClientI>;
  accountContext?: Partial<AccountContextI>;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  accountContext,
  apiContext = { apiClient: { get: () => undefined as any } },
  children
}: MockAppContextProviderProps) {
  const DEFAULT_MOCK_ACCOUNT_CONTEXT: AccountContextI = useMemo(
    () => ({
      authenticated: true,
      groupNames: ["aafc", "cnc"],
      initialized: true,
      login: noop,
      logout: noop,
      roles: ["user"],
      token: "test-token",
      username: "test-user",
      isAdmin: false
    }),
    []
  );

  const DEFAULT_API_CONTEXT_VALUE = useMemo(
    () =>
      new ApiClientImpl({
        newId: () => "00000000-0000-0000-0000-000000000000"
      }),
    []
  );

  const apiContextWithWarnings = {
    ...apiContext,
    // Add a warning when bulkGet doesn't return anything in a test:
    bulkGet: async (paths: string[], options) => {
      const resources = await apiContext?.bulkGet?.(paths, options);
      if (!resources) {
        console.warn("No response returned for bulkGet paths: ", paths);
      } else {
        resources.forEach((resource, index) => {
          if (!resource) {
            console.warn("No value returned for bulkGet path: ", paths[index]);
          }
        });
      }
      return resources;
    }
  };

  const modalWrapperRef = useRef<HTMLDivElement>(null);

  return (
    <SWRConfig
      // Reset SWR cache between tests.
      value={{ provider: () => new Map() }}
    >
      <AccountProvider
        value={{ ...DEFAULT_MOCK_ACCOUNT_CONTEXT, ...accountContext }}
      >
        <ApiClientProvider
          value={merge({}, DEFAULT_API_CONTEXT_VALUE, apiContextWithWarnings)}
        >
          <FileUploadProviderImpl>
            <DinaIntlProvider>
              <DndProvider backend={HTML5Backend}>
                <div ref={modalWrapperRef}>
                  <ModalProvider appElement={modalWrapperRef.current}>
                    {children}
                  </ModalProvider>
                </div>
              </DndProvider>
            </DinaIntlProvider>
          </FileUploadProviderImpl>
        </ApiClientProvider>
      </AccountProvider>
    </SWRConfig>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}
