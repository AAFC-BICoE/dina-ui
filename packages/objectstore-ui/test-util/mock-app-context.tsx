import {
  AccountContextI,
  AccountProvider,
  ApiClientContextI,
  AuthenticatedApiClientProvider,
  createContextValue,
  ModalProvider
} from "common-ui";
import { mount } from "enzyme";
import { merge, noop } from "lodash";
import { ObjectStoreIntlProvider } from "../intl/objectstore-intl";

interface MockAppContextProviderProps {
  apiContext?: Partial<ApiClientContextI>;
  accountContext?: Partial<AccountContextI>;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  accountContext,
  apiContext,
  children
}: MockAppContextProviderProps) {
  return (
    <AccountProvider
      value={{ ...DEFAULT_MOCK_ACCOUNT_CONTEXT, ...accountContext }}
    >
      <AuthenticatedApiClientProvider
        apiContext={merge({}, DEFAULT_API_CONTEXT_VALUE, apiContext)}
      >
        <ObjectStoreIntlProvider>
          <ModalProvider appElement={document.querySelector("body")}>
            {children}
          </ModalProvider>
        </ObjectStoreIntlProvider>
      </AuthenticatedApiClientProvider>
    </AccountProvider>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}

const DEFAULT_MOCK_ACCOUNT_CONTEXT: AccountContextI = {
  authenticated: true,
  groups: ["/aafc", "cnc"],
  initialized: true,
  login: noop,
  logout: noop,
  token: "test-token",
  username: "test-user"
};

const DEFAULT_API_CONTEXT_VALUE = createContextValue({
  getTempIdGenerator: () => () => "00000000-0000-0000-0000-000000000000"
});
