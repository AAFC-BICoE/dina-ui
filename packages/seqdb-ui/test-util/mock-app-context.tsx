import {
  ApiClientContext,
  ApiClientContextI,
  createContextValue
} from "common-ui";
import { mount } from "enzyme";
import { SeqdbIntlProvider } from "../intl/seqdb-intl";

interface MockAppContextProviderProps {
  apiContext?: ApiClientContextI;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  apiContext,
  children
}: MockAppContextProviderProps) {
  return (
    <ApiClientContext.Provider value={apiContext || createContextValue()}>
      <SeqdbIntlProvider>{children}</SeqdbIntlProvider>
    </ApiClientContext.Provider>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}
