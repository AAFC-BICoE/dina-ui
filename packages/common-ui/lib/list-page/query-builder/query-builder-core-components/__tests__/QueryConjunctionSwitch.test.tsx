import { mountWithAppContext } from "common-ui/lib/test-util/mock-app-context";
import { useState } from "react";
import { QueryConjunctionSwitch } from "../QueryConjunctionSwitch";

describe("QueryConjunctionSwitch component", () => {
  test("Snapshot Test", async () => {
    const andToggled = mountWithAppContext(
      <>
        <QueryConjunctionSwitch currentConjunction="AND" />
      </>
    );
    expect(andToggled.find(QueryConjunctionSwitch).debug()).toMatchSnapshot();

    const orToggled = mountWithAppContext(
      <>
        <QueryConjunctionSwitch currentConjunction="OR" />
      </>
    );
    expect(orToggled.find(QueryConjunctionSwitch).debug()).toMatchSnapshot();
  });

  test("Toggle between the options", async () => {
    function TestToggle() {
      const [conjunction, setConjunction] = useState<string>("AND");

      return (
        <QueryConjunctionSwitch
          currentConjunction={conjunction}
          setConjunction={(newConjunction) => setConjunction(newConjunction)}
        />
      );
    }

    const toggle = mountWithAppContext(
      <>
        <TestToggle />
      </>
    );

    // Expect the initial state.
    expect(
      toggle.find(QueryConjunctionSwitch).prop("currentConjunction")
    ).toEqual("AND");

    // Click the toggle to "OR".
    toggle.find("button").at(1).simulate("click");
    expect(
      toggle.find(QueryConjunctionSwitch).prop("currentConjunction")
    ).toEqual("OR");

    // Switch back to "AND"
    toggle.find("button").at(0).simulate("click");
    expect(
      toggle.find(QueryConjunctionSwitch).prop("currentConjunction")
    ).toEqual("AND");
  });
});
