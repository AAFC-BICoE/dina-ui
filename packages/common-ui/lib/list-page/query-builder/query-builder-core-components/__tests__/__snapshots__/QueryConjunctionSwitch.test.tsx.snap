// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryConjunctionSwitch component Snapshot Test 1`] = `
"<QueryConjunctionSwitch currentConjunction=\\"AND\\">
  <div className=\\"toggleGroup\\">
    <Button className=\\"toggleButton activeToggle\\" onClick={[Function: onClick]} variant=\\"primary\\" active={false} disabled={false}>
      <button onClick={[Function: onClick]} disabled={false} type=\\"button\\" className=\\"toggleButton activeToggle btn btn-primary\\">
        <FormattedMessage id=\\"queryBuilder_conjunction_and\\">
          <FormattedMessage defaultMessage=\\" \\" id=\\"queryBuilder_conjunction_and\\">
             
          </FormattedMessage>
        </FormattedMessage>
      </button>
    </Button>
    <Button className=\\"toggleButton\\" onClick={[Function: onClick]} variant=\\"primary\\" active={false} disabled={false}>
      <button onClick={[Function: onClick]} disabled={false} type=\\"button\\" className=\\"toggleButton btn btn-primary\\">
        <FormattedMessage id=\\"queryBuilder_conjunction_or\\">
          <FormattedMessage defaultMessage=\\" \\" id=\\"queryBuilder_conjunction_or\\">
             
          </FormattedMessage>
        </FormattedMessage>
      </button>
    </Button>
  </div>
</QueryConjunctionSwitch>"
`;

exports[`QueryConjunctionSwitch component Snapshot Test 2`] = `
"<QueryConjunctionSwitch currentConjunction=\\"OR\\">
  <div className=\\"toggleGroup\\">
    <Button className=\\"toggleButton\\" onClick={[Function: onClick]} variant=\\"primary\\" active={false} disabled={false}>
      <button onClick={[Function: onClick]} disabled={false} type=\\"button\\" className=\\"toggleButton btn btn-primary\\">
        <FormattedMessage id=\\"queryBuilder_conjunction_and\\">
          <FormattedMessage defaultMessage=\\" \\" id=\\"queryBuilder_conjunction_and\\">
             
          </FormattedMessage>
        </FormattedMessage>
      </button>
    </Button>
    <Button className=\\"toggleButton activeToggle\\" onClick={[Function: onClick]} variant=\\"primary\\" active={false} disabled={false}>
      <button onClick={[Function: onClick]} disabled={false} type=\\"button\\" className=\\"toggleButton activeToggle btn btn-primary\\">
        <FormattedMessage id=\\"queryBuilder_conjunction_or\\">
          <FormattedMessage defaultMessage=\\" \\" id=\\"queryBuilder_conjunction_or\\">
             
          </FormattedMessage>
        </FormattedMessage>
      </button>
    </Button>
  </div>
</QueryConjunctionSwitch>"
`;
